### PR TITLE
Added gitignore-file for Beckhoff TwinCAT3.

### PR DESCRIPTION
**Pull request by sagatowski**
**Provide a link to the application or project’s homepage:**
https://www.beckhoff.com/english.asp?twincat/twincat-3.htm
TwinCAT3 is a realtime kernel and development environment for PLCs (programmable logic controller). The IDE is based on Visual Studio, which TwinCAT is integrated into. It's one of the more popular development environments for industrial automation.

**Provide links to documentation:**
The official one:
https://download.beckhoff.com/download/document/automation/twincat3/TC3_SourceControl_EN.pdf
This only covers part of the files (specifying that TMC and TPY-files are not necessary to version control). Unfortunately, even Beckhoff (the manufacturer of TwinCAT) have failed to properly document which files should be version controlled and not.

Regarding the other files included in the gitignore-file of the pull request:
- *.tclrs - These are license response files, basically license files that are residing on the target PLC device
- *.compiled-library and *.library - which are libraries of which the project depends on. These are put here when they are referenced in the project, but the actual libraries needed for development are installed in "C:\TwinCAT\3.1\Components\Plc\Managed Libraries", which need to be installed there prior of compilation. The ones residing in the project must not be version controlled.
- *.compileinfo - Self explaining, file created everytime a compilation is made. Beckhoff documentation says: "The compile information was created during the last download of the PLC project and stored in a file".
- _Boot/ - Folder generated when project is compiled with all the binaries. This is what is transferred to the actual target device (PLC) for execution.
- _CompileInfo/ - Self explaining. Folder created during compilation. Holding the *.compileinfo-files.
- _Libraries/ - Folder holding the *.compiled-library and *.library in the local development folder. 

There are various github repos for TwinCAT that use this list for gitignore, one being TcUnit, a unit testing framework for TwinCAT3:
https://github.com/tcunit/TcUnit/blob/master/.gitignore

Note that because TwinCAT is using visual studio as development environment (it integrates itself into VS during installation), all visual studio files that are normally ignored shall be ignored as well (therefore the comment # Recommended: VisualStudio.gitignore).

Update 2019-12-05: I've published a blog post that goes through these files and the reason for having a gitignore-file for TwinCAT3 in a much greater detail now which is accessible here:
https://alltwincat.com/2019/12/02/gitignore-for-twincat/

**Explain why you’re making a change:**
Unfortunately version control by modern VCS is yet quite uncommon in industrial automation, and therefore proper practices and documentation around it have not yet been fully defined. It's quite common within the world of automation to do version control by simply taking a complete project folder and copying it to a new folder, adding the new date and calling that version control. Now that more and more automation engineers are starting to do version control more properly, and also starting to use GIT, it makes sense to create this file.

**If this is a new template:**
Link to application or projects homepage:
This would be TcUnit, which is a popular unit testing framework for TwinCAT3:
https://tcunit.org/
https://github.com/tcunit

### CODE REVIEW

**Comment by sagatowski** [comment]
@shiftkey - it's been two months since the creation of this PR and I just wanted to double-check if there is something expected from me that I need to do for this to progress? I tried to document the reason for this addition as well as I could, but maybe there is still some information missing that you expect?

**Comment by sagatowski** [comment]
@shiftkey , just doing a ping on this one as it's now been 5 months since the creation of this ticket... :-P

**Comment by massimo-pizzardi** [comment]
I think that is a good idea to add this line:
*.tsproj.bk*
 these are backup file of *.tsproj

**Comment by sagatowski** [comment]
> I think that is a good idea to add this line:
> _.tsproj.bk_
> these are backup file of *.tsproj

I've already included *.tsproj.bak... which I assume you meant?

**Comment by massimo-pizzardi** [comment]
.tsproj.bk1
.tsproj.bk2
.tsproj.bk3
thise files are backup of file .tsproj.bak

**Comment by rudischilder** [comment]
### TwinCAT HMI

Assuming this .gitignore should be applicable to all TwinCAT modules, you can add the following:

```markdown
## Ignore TwinCAT HMI temporary files, build results, and
## files generated by popular TwinCAT HMI add-ons.
.engineering_servers
liveview_*
*.cache
*.db-shm
*.db-wal
*.pid
tchmipublish.journal.json
```

These come from a file called `.tfignore`, which is generated upon adding a new TwinCAT HMI project to a solution, somewhere in your working tree. 

Initially I thought that anything under `\Server\Extensions` should also be ignored, but that is up for debate. For some reason, the underlying .json files are often seen as changed, while in fact they are not. This leads me to believe they were generated. However they do contain information necessary for publishing your HMI project to a HMI server. For now I myself do not ignore them. Instead I undo any changes made to them, unless I can explain those changes. 

### Backup files
I noted that visual studio/shell did not always recognize backup files, so i simply added `*.bak` as well. 

It should be noted that PLC projects come with the `.tcpproj` extension, and thus also the `.tcpproj.bak` should be added to the gitignore file. 

### Final note
I do not like that the .tmc file must be generated on build, yet saved to the repo. I hope Beckhoff could split any necessary configuration/generation in future versions.

**Comment by HolgerJeromin** [comment]
@rudischilder 
In the TwinCAT HMI 1.12 beta there are the following further entries added to its `.gitignore`:

```
.hmiframework/
.hmipkgs/*-*-*-*/
```

**Comment by sagatowski** [comment]
@rudischilder & @HolgerJeromin, when creating a TwinCAT-HMI project, a .gitignore-file is created automatically, thus should not be included here.

**Review by JonatanBang — APPROVED** [review]

**Comment by isherman** [comment]
@sagatowski thank you for this.
Should the `_Deployment` directory be ignored as well? The `_Repository` directory?

**Comment by sagatowski** [comment]
> @sagatowski thank you for this.
> Should the `_Deployment` directory be ignored as well? The `_Repository` directory?

@isherman I haven't managed to reproduce the _Deployment directory. Under what conditions is this created?

**Review by SankariNL — APPROVED** [review]

**Review by kryskool — APPROVED** [review]
Ok for me
Use similar file on my own repository

**Comment by StEvUgnIn** [comment]
Hi,
Any update? This work seems alright to me.

**Comment by sagatowski** [comment]
> Hi,
> Any update? This work seems alright to me.

GitHub are not in a hurry with approving this PR it seems.

@shiftkey It's been almost two years since this PR was created. How much (approx.) will it take before it is approved? If there is some information missing, please let me know.

File: TwinCAT.gitignore
Comment (StEvUgnIn) [inline TwinCAT.gitignore]:
```suggestion
```

File: TwinCAT.gitignore
Comment (StEvUgnIn) [inline TwinCAT.gitignore]:
```suggestion
```

**Review by StEvUgnIn — CHANGES_REQUESTED** [review]
I would drop these two lines

File: TwinCAT.gitignore
Comment (sagatowski) [reply TwinCAT.gitignore]: You're watching an outdated version. The new version is correctly writing "TwinCAT 3".

Comment (StEvUgnIn) [reply TwinCAT.gitignore]: Rebase the branch

File: TwinCAT3.gitignore
Comment (StEvUgnIn) [inline TwinCAT3.gitignore:4]:
```suggestion
```

File: TwinCAT3.gitignore
Comment (StEvUgnIn) [inline TwinCAT3.gitignore:3]:
```suggestion
```

**Review by StEvUgnIn — CHANGES_REQUESTED** [review]
I will do someting nicer.

File: TwinCAT3.gitignore
Comment (StEvUgnIn) [inline TwinCAT3.gitignore:1]:
```suggestion 
# gitignore template for TwinCAT3 - https://www.beckhoff.com/twincat3/
```

File: TwinCAT3.gitignore
Comment (StEvUgnIn) [inline TwinCAT3.gitignore:3]:
```suggest
```

Comment (StEvUgnIn) [inline TwinCAT3.gitignore:3]:
```suggestion
```

File: TwinCAT3.gitignore
Comment (StEvUgnIn) [inline TwinCAT3.gitignore:4]:
```suggestion
```

File: TwinCAT3.gitignore
Comment (StEvUgnIn) [inline TwinCAT3.gitignore]: newline missing
```suggestion
_Libraries/

```

**Comment by isherman** [comment]
> @isherman I haven't managed to reproduce the `_Deployment` directory. Under what conditions is this created?

I haven't tried to isolate the exact condition that creates it, but my project includes a TwinCAT C++ module and TwinSAFE configuration.

**Review by StEvUgnIn — CHANGES_REQUESTED** [review]
Please make the proper changes

File: TwinCAT3.gitignore
Comment (StEvUgnIn) [inline TwinCAT3.gitignore:25]: missing newline character before EOF
```suggestion
_ModuleInstall/

```

File: TwinCAT.gitignore
Comment (StEvUgnIn) [inline TwinCAT.gitignore]: Commiting TMC is up to the user
```suggestion
#*.tmc
```

File: TwinCAT3.gitignore
Comment (StEvUgnIn) [inline TwinCAT3.gitignore:11]:
```suggestion
# uncomment if you are using TwinCAT3 with a general purpose programming language (C++, C#, ...)
```

File: TwinCAT3.gitignore
Comment (StEvUgnIn) [inline TwinCAT3.gitignore:12]:
```suggestion
```

**Review by StEvUgnIn — CHANGES_REQUESTED** [review]
passing concise comment

File: TwinCAT3.gitignore
Comment (StEvUgnIn) [inline TwinCAT3.gitignore:14]:
```suggestion
#*.tmc
```

File: TwinCAT3.gitignore
Comment (StEvUgnIn) [inline TwinCAT3.gitignore:15]: Commiting TMC is up to the user
```suggestion
#*.tmcRefac
```

**Comment by mazid12345** [comment]
[*] unc0ver is not for sale
[*] If you purchased unc0ver, please report the seller
[*] Get unc0ver for free at https://unc0ver.dev
[*] Configured to share anonymous OS crash logs
[*] Machine Name: iPhone9,4
[*] Model Name: D111AP
- [*] Kernel Version: Darwin Kernel Version 20.1.0: Fri Oct 30 00:34:15 PDT 2020; root:xnu-7195.42.3~1/RELEASE[]()

**Review by StEvUgnIn — CHANGES_REQUESTED** [review]
What do you think?

File: TwinCAT3.gitignore
Comment (mazid12345) [reply TwinCAT3.gitignore:14]: > ```suggestion
> 
> #*.
> 
> ```

File: TwinCAT.gitignore
Comment (sagatowski) [reply TwinCAT.gitignore]: No it's not. Since 4018 Beckhoff specifically says tmc should be ignored (other for the exception of C++, which is also explicitly mentioned).

File: TwinCAT3.gitignore
Comment (sagatowski) [reply TwinCAT3.gitignore:14]: See previous comment.

File: TwinCAT3.gitignore
Comment (sagatowski) [reply TwinCAT3.gitignore:15]: See previous comment about TMC.

File: TwinCAT3.gitignore
Comment (sagatowski) [reply TwinCAT3.gitignore:11]: General purpose languages have nothing to do with TMC-files, other than specifically IEC61131-3 (PLC languages) and specifically C++.

**Comment by sagatowski** [comment]
> [_] unc0ver is not for sale [_] If you purchased unc0ver, please report the seller
> [_] Get unc0ver for free at https://unc0ver.dev [_] Configured to share anonymous OS crash logs
> [_] Machine Name: iPhone9,4 [_] Model Name: D111AP
> 
> * [*] Kernel Version: Darwin Kernel Version 20.1.0: Fri Oct 30 00:34:15 PDT 2020; root:xnu-7195.42.3~1/RELEASE

Great, now we've got SPAM-bots attacking this PR too. I hope this PR can be accepted & closed.

**Comment by oscarbailey-tc** [comment]
Just to add, I'd like this PR merged too - used the .gitignore yesterday for a TC3 project, thanks @sagatowski

**Comment by StEvUgnIn** [comment]
> Hey folks, sorry for the long wait. LGTM

My suggestions were not taken into account..

**Pull request merged**
